addons/website_sale_delivery/static/src/js/website_sale_delivery.js

```diff
@@ -164,5 +203,15 @@ publicWidget.registry.websiteSaleDelivery = publicWidget.Widget.extend({
             $providerRestricted.hide().attr('disabled', true);
         }
     },
+    /**
+     * @private
+     * @param {Event} ev
+     */
+    _onPaymentClick: function (ev) {
+        const $payButton = $('button[name="o_payment_submit_button"]');
+        if (!this._areSelectedPaymentAndCarrierCompatible()) {
+           $payButton.prop('disabled', true);
+        }
```

Comment:
Do you need to disable the button at both points ? here and in the `_handleCarrierUpdateResult ` ? If the function is called each time a carrier is selected it should be enough to disable the button once.

Reply:
There are two state we need to handle:
- When we have COD and UPS already selected and we change the carrier, this is handled by the `_handCarrierUpdateResult` as you mentioned, since we are clicking on a carrier.
- When we have <not_cod_provider> and <not_ups_carrier>, and we change the prodider to COD, then we need to disable the button (since COD can't be paired with no other carrier than UPS) and this is handle by this event hook `_onPaymentClick`.